### PR TITLE
[refactor] Remove fieldname argument from the field validators

### PR DESCRIPTION
**Pull request by jjotero**
Field validators used the `fieldname` argument in the `__init__` method to get the class variable name they were assigned to. That led to redundant typing of the attribute's name twice:
```python
cc = fields.TypeField('cc', str)
```
This field validators now have the `__set_name__` method implemented, which automatically imports the class variable name:
```python
cc = fields.TypeField(str)
```

### CODE REVIEW

File: reframe/core/containers.py [hunk 1]
```diff
@@ -32,14 +32,14 @@ class ContainerPlatform(abc.ABC):
     #:
     #: :type: :class:`list[tuple[str, str]]`
     #: :default: ``[]``
-    mount_points = fields.TypedField('mount_points',
-                                     typ.List[typ.Tuple[str, str]])
+    mount_points = fields.TypedField(
+        typ.List[typ.Tuple[str, str]])
```
Comment (teojgo) [inline reframe/core/containers.py:36]:
```suggestion
    mount_points = fields.TypedField(typ.List[typ.Tuple[str, str]])
```

File: reframe/core/schedulers/__init__.py [hunk 2]
```diff
@@ -123,24 +123,24 @@ class Job(jsonext.JSONSerializable):
...
-                                          int,  type(None))
-    use_smt = fields.TypedField('use_smt', bool,  type(None))
-    time_limit = fields.TimerField('time_limit', type(None))
+    num_tasks = fields.TypedField(int)
+    num_tasks_per_node = fields.TypedField(
+        int,  type(None))
+    num_tasks_per_core = fields.TypedField(
+        int,  type(None))
+    num_tasks_per_socket = fields.TypedField(
+        int,  type(None))
```
Comment (teojgo) [inline reframe/core/schedulers/__init__.py:132]:
```suggestion
    num_tasks_per_socket = fields.TypedField(int,  type(None))
```

File: reframe/core/schedulers/__init__.py [hunk 3]
```diff
@@ -123,24 +123,24 @@ class Job(jsonext.JSONSerializable):
...
+
+    num_cpus_per_task = fields.TypedField(
+        int,  type(None))
```
Comment (teojgo) [inline reframe/core/schedulers/__init__.py:135]:
```suggestion
    num_cpus_per_task = fields.TypedField(int,  type(None))
```

File: reframe/core/schedulers/__init__.py [hunk 4]
```diff
@@ -123,24 +123,24 @@ class Job(jsonext.JSONSerializable):
...
+    use_smt = fields.TypedField(bool,  type(None))
```
Comment (teojgo) [inline reframe/core/schedulers/__init__.py:136]:
```suggestion
    use_smt = fields.TypedField(bool, type(None))
```

File: reframe/core/pipeline.py [hunk 5]
```diff
@@ -446,42 +442,42 @@ def pipeline_hooks(cls):
...
     #: Specify whether this tests needs simultaneous multithreading enabled.
     #:
     #: Ignored if :class:`None`.
     #:
     #: :type: boolean or :class:`None`
     #: :default: :class:`None`
-    use_multithreading = fields.TypedField('use_multithreading',
-                                           bool, type(None))
+    use_multithreading = fields.TypedField(
+        bool, type(None))
```
Comment (teojgo) [inline reframe/core/pipeline.py:480]:
```suggestion
    use_multithreading = fields.TypedField(bool, type(None))
```

File: reframe/core/pipeline.py [hunk 6]
```diff
@@ -446,42 +442,42 @@ def pipeline_hooks(cls):
...
     #: Number of tasks per socket required by this test.
     #:
     #: Ignored if :class:`None`.
     #:
     #: :type: integral or :class:`None`
     #: :default: :class:`None`
-    num_tasks_per_socket = fields.TypedField('num_tasks_per_socket',
-                                             int, type(None))
+    num_tasks_per_socket = fields.TypedField(
+        int, type(None))
```
Comment (teojgo) [inline reframe/core/pipeline.py:471]:
```suggestion
    num_tasks_per_socket = fields.TypedField(int, type(None))
```

File: reframe/core/pipeline.py [hunk 7]
```diff
@@ -446,42 +442,42 @@ def pipeline_hooks(cls):
...
     #: Number of tasks per core required by this test.
     #:
     #: Ignored if :class:`None`.
     #:
     #: :type: integral or :class:`None`
     #: :default: :class:`None`
-    num_tasks_per_core = fields.TypedField('num_tasks_per_core',
-                                           int, type(None))
+    num_tasks_per_core = fields.TypedField(
+        int, type(None))
```
Comment (teojgo) [inline reframe/core/pipeline.py:462]:
```suggestion
    num_tasks_per_core = fields.TypedField(int, type(None))
```

File: reframe/core/containers.py (hunk 1 above)
Comment (teojgo) [inline reframe/core/containers.py:36]:
```suggestion
    mount_points = fields.TypedField(typ.List[typ.Tuple[str, str]])
```

**Review by teojgo — CHANGES_REQUESTED** [review]
Fix also the `unittests/test_fields.py` to use the new field syntax.

**Comment by jjotero** [comment]
Jenkins CI passed after a retry.

**Comment by teojgo** [comment]
@jenkins-cscs retry all

**Review by teojgo — APPROVED** [review]

File: reframe/core/fields.py [hunk 8]
```diff
@@ -42,6 +42,9 @@ def __set__(self, obj, value):
 class ForwardField:
```
Comment (vkarak) [inline reframe/core/fields.py:42]: I suggest removing completely the `ForwardField`. It's not used anywhere and it was a more verbose way for forwarding attribute access to a backend, which is now done much more easily with `__getattr__()`.

File: unittests/test_fields.py [hunk 9]
```diff
@@ -26,7 +26,7 @@ class FieldTester:
 
 def test_copy_on_write_field():
     class FieldTester:
-        cow = fields.CopyOnWriteField('cow')
+        cow = fields.CopyOnWriteField()
```
Comment (vkarak) [inline unittests/test_fields.py:29]: You can also remove this field completely. It's not used anywhere.

File: unittests/test_fields.py [hunk 10]
```diff
@@ -206,7 +205,7 @@ def __init__(self):
 
 def test_absolute_path_field():
     class FieldTester:
-        value = fields.AbsolutePathField('value', type(None))
+        value = fields.AbsolutePathField(type(None))
```
Comment (vkarak) [inline unittests/test_fields.py:208]: Same here.

**Review by vkarak — CHANGES_REQUESTED** [review]
Since you touched the fields, let's take the opportunity to remove some of them that are stale. Other than that, this is a very cool and useful feature that I wasn't aware of!

File: reframe/core/fields.py (hunk 8 above)
Comment (jjotero) [reply reframe/core/fields.py]: Done

File: unittests/test_fields.py (hunk 10 above)
Comment (jjotero) [reply unittests/test_fields.py]: Done

File: unittests/test_fields.py (hunk 9 above)
Comment (jjotero) [reply unittests/test_fields.py]: Done

**Review by vkarak — APPROVED** [review]
lgtm

**Comment by vkarak** [comment]
@jjotero Can you also fix the unused imports that the CI complains about?

**Review by vkarak — APPROVED** [review]
lgtm

**Comment by jjotero** [comment]
@jenkins-cscs retry all

**Pull request merged**